fendermint/testing/materializer/src/docker/relayer.rs

```diff
@@ -1,20 +1,129 @@
 // Copyright 2022-2024 Protocol Labs
 // SPDX-License-Identifier: Apache-2.0, MIT
 
-use std::fmt::Display;
+use std::{fmt::Display, path::Path};
 
-use crate::RelayerName;
+use anyhow::Context;
+use bollard::Docker;
 
-use super::DockerConstruct;
+use crate::{
+    docker::{
+        runner::{split_cmd, DockerRunner},
+        user_id, FENDERMINT_IMAGE,
+    },
+    manifest::EnvMap,
+    materials::{DefaultAccount, DefaultSubnet},
+    RelayerName, ResourceHash, TestnetResource,
+};
+
+use super::{container::DockerContainer, dropper::DropChute, network::NetworkName, DropPolicy};
 
-/// TODO: Dockerize the relayer
 pub struct DockerRelayer {
     relayer_name: RelayerName,
-    container: DockerConstruct,
+    relayer: DockerContainer,
 }
 
 impl Display for DockerRelayer {
     fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
         Display::fmt(&self.relayer_name, f)
     }
 }
+
+impl DockerRelayer {
+    /// Get or create the relayer container.
+    ///
+    /// This assumes that the submitter and the involved parent and child
+    /// subnets have been added to the `ipc-cli` config.
+    #[allow(clippy::too_many_arguments)]
+    pub async fn get_or_create<'a>(
+        root: impl AsRef<Path>,
+        docker: Docker,
+        dropper: DropChute,
+        drop_policy: &DropPolicy,
+        relayer_name: &RelayerName,
+        subnet: &DefaultSubnet,
+        submitter: &DefaultAccount,
+        network_name: Option<NetworkName>,
+        env: &EnvMap,
+    ) -> anyhow::Result<Self> {
+        let container_name = container_name(relayer_name);
+
+        // If the container exists, return it.
+        if let Some(relayer) = DockerContainer::get(
+            docker.clone(),
+            dropper.clone(),
+            drop_policy,
+            container_name.clone(),
+        )
+        .await?
+        {
+            return Ok(Self {
+                relayer_name: relayer_name.clone(),
+                relayer,
+            });
+        }
+
+        // We'll need to mount the IPC configuration for the relayer.
+        let ipc_dir = root.as_ref().join(subnet.name.testnet()).join("ipc");
+
+        let user = user_id(&ipc_dir)?;
+
+        // The CLI only logs to the output. Its log level can be configured with the general env vars.
+        let volumes = vec![(ipc_dir, "/fendermint/.ipc")];
+
+        let creator = DockerRunner::new(
+            docker,
+            dropper,
+            drop_policy.clone(),
+            relayer_name.clone(),
+            user,
+            FENDERMINT_IMAGE,
```

Comment:
Shouldn't we make the image configurable?

Reply:
We should, yes. I wanted to do it in a followup PR, just not sure where to put such configuration. It doesn't belong in the manifest. Maybe an env var, or just a general config file that that the materializer interprets according to how it's implemented.